### PR TITLE
Fix RSP DropZone's filled message announcement on different platforms

### PR DESCRIPTION
**Pull request by yihuiliao**
**TLDR:**

Specifically for Chrome + VO, during a virtual drag, VO will not announce any of the labels. However, Chrome + VO is happy to announce labels during a virtual drag so long as there isn't both an `aria-label` and `aria-labelledby` on the element that is receiving focus. As a result, the solution is to add a `div` to the `<VisuallyHidden>` component that receives the`aria-label` prop and is pointed to by the `aria-labelledby`. We can combine all the different elements that make up drop zone’s accessible name and pass it to useLabels as a single `aria-labelledby`.

_____ 

**The Longer Version:** 

Originally, the issue was that some platforms were not announcing the filled message that says “Drop File to Replace”. 

Upon further investigation, it seemed that there was a larger issue, specifically for Chrome + VO. In short, during a virtual drag, it wasn’t announcing any labels and would skip to the `aria-describedby`. At least for the other platforms, if it wasn’t announcing the filled message, it was at least announcing the labels. 
Funny enough, Chrome + VO has no problem announcing the labels when you were not in a virtual drag and just focusing on the visually hidden button. This problem is specific to virtual dragging. 

We haven’t pinpointed what the exact issue is, especially since this seems to be specific to Chrome + VO. However, we do have a workaround that seems acceptable enough. It hinges on the fact that Chrome + VO is capable of announcing labels during a virtual drag BUT ONLY under a specific condition which is that the visually hidden button only have either an `aria-label` OR an `aria-labelledby` but not together. 

The solution is to add a `div` to the `<VisuallyHidden>` component that recieves the`aria-label` prop and is pointed to by the `aria-labelledby`. We can combine all the different elements that make up drop zone’s accessible name and pass it to useLabels as a single `aria-labelledby`. Chrome + VO is happy to announce this. 

I will mention that Chrome + VO is still rather finicky and doesn’t always announce the filled message (though it will at least announce the other labels). Also, depending on how quickly you decide to drag and drop, it might just skip announcing the label entirely. I recommend when testing to go slowly and to give the screenreader some time to talk.

Things that we tried while debugging:
1. Creating a `<Draggable />` and `<DropTarget />` component that used the hooks to see if the same issue still applied. The answer is that it did. Seems to be an issue with drag and drop and not specifically `<DropZone />`. Here’s the [codesandbox](https://codesandbox.io/s/drag-and-drop-investigation-7j7nkl)
2. Created a minimal reproduction with no drag and drop to see if there was an issue with combining aria-labels and aria-labelledby’s on the same element. Here’s the [codesandbox](https://codesandbox.io/s/more-aria-label-dropzone-testing-spy9j8)
3. Clear out the aria-describedby that is passed by dropButtonProps by setting it to null. Reads out only ‘button’ on a virtual drag. 
4. Commented out the announcements from DragManager. Labels were still not being announced. 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

When the dropzone is filled, the label "Drop file to replace" will always be announced.

1. Turn on screenreader
2. Go to "WithDraggable"
3. Tab to focus on the visually hidden button. Should say "DropZone Drag and drop here"
4. Enter a virtual drag, when dragging over the DropZone the first time, it should say "DropZone Drag and drop here". Drag over the DropZone a second time, it should say "DropZone Drag and drop here Drop file to replace"
5. When the dropzone is in a filled state, focusing on the visually hidden button should announce "DropZone Drag and drop here Drop file to replace"
7. Go to "DropZone Example With Draggable Object" in RAC section of Storybook
8. Tab to focus on the visually hidden button. Should say "DropZone DropZone Area"
9. Enter a virtual drag, when dragging over the DropZone, it should say "DropZone DropZone Area"

### CODE REVIEW

File: packages/react-aria-components/src/DropZone.tsx [hunk 1]
```diff
@@ -53,12 +54,16 @@ function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
...
   let stringFormatter = useLocalizedStringFormatter(intlMessages);
+  let isVirtualDrag = isVirtualDragging();
 
   let textId = useSlotId();
+  let dropzoneId = useSlotId();
   let ariaLabel = props['aria-label'] || stringFormatter.format('dropzoneLabel');
-  let messageId = (isDropTarget && props['aria-labelledby']) ? props['aria-labelledby'] : null;
-  let ariaLabelledby = [textId, messageId].filter(Boolean).join(' ');
-  let labelProps = useLabels({'aria-label': ariaLabel, 'aria-labelledby': ariaLabelledby});
+  let messageId = (isVirtualDrag && props['aria-labelledby']) ? props['aria-labelledby'] : null;
```
Comment (yihuiliao) [inline packages/react-aria-components/src/DropZone.tsx:62]: I changed this from `isDropTarget` to `isVirtualDrag` because it wasn't announcing on iOS(). It seemed that because the `aria-labelledby` only pointed to the filled message when `isDropTarget=true` that it was rendering a bit too late for the screenreader to announce. As a result, I changed it to `isVirtualDrag` so that the filled message's `aria-labelledby` is added earlier so the screenreader will announce it. Not sure if this is something we want to include though.

File: packages/react-aria-components/src/DropZone.tsx [hunk 2]
```diff
@@ -95,6 +100,10 @@ function DropZone(props: DropZoneProps, ref: ForwardedRef<HTMLDivElement>) {
         data-focus-visible={isFocusVisible || undefined}
         data-drop-target={isDropTarget || undefined} >
         <VisuallyHidden>
+          {/* Added as a workaround for a Chrome + VO bug where it will not announce the aria label */}
+          <div id={dropzoneId} aria-hidden="true">
```
Comment (yihuiliao) [inline packages/react-aria-components/src/DropZone.tsx:104]: Originally, I wanted to do something like `<div id={dropzoneId} aria-label={ariaLabel} aria-hidden="true></div>` but Chrome + VO wouldn't announce the label (though it seemed fine on other platforms). I created a [codesandbox](https://codesandbox.io/s/button-labels-investigation-lz5nwf?file=/src/App.js) trying various combinations to see how Chrome + VO was announcing it

**Review by snowystinger — APPROVED** [review]

File: packages/react-aria-components/src/DropZone.tsx (hunk 1 above)
Comment (LFDanLu) [reply packages/react-aria-components/src/DropZone.tsx:62]: Actually, should we even have a conditional on `messageId` here? Don't we want to apply the user provided `aria-labelledby` regardless if it is a virtual drag or not?

Comment (yihuiliao) [reply packages/react-aria-components/src/DropZone.tsx:62]: My thought process was that we only wanted to announce the filled message when it was being displayed which is during a drag. Not sure which is correct though

Comment (reidbarber) [reply packages/react-aria-components/src/DropZone.tsx:62]: I think having it on there regardless of if its a virtual drag is the way to go. I think if the user is providing it, they would expect it to be applied.

Comment (LFDanLu) [reply packages/react-aria-components/src/DropZone.tsx:62]: Thats a RSP specific implementation detail though right? IMO for RAC Dropzone we should always have the user provided `aria-labelledby` applied since it may be used to label the dropzone in lieu of providing some visible Text as described in our [docs](https://reactspectrum.blob.core.windows.net/reactspectrum/de8064731dc91737cb3fad55de0d4716d21c37d7/docs/react-aria/DropZone.html#anatomy).

Comment (LFDanLu) [reply packages/react-aria-components/src/DropZone.tsx:62]: If we don't want the RSP Dropzone to announce the virtual drop specific message, can we access `let isVirtualDrag = isVirtualDragging();` in the RSP Dropzone? If we can, then you can always provide the user's `aria-labelledby` and only append our own `messageId` for the `isFilled` message to it if `isVirtualDrag` is true

**Review by snowystinger — DISMISSED** [review]

File: packages/@react-spectrum/dropzone/src/DropZone.tsx [hunk 3]
```diff
@@ -35,12 +37,17 @@ function DropZone(props: SpectrumDropZoneProps, ref: DOMRef<HTMLDivElement>) {
   let domRef = useDOMRef(ref);
   let messageId = useId();
   let stringFormatter = useLocalizedStringFormatter(intlMessages);
+  let isVirtualDrag = isVirtualDragging();
```
Comment (yihuiliao) [inline packages/@react-spectrum/dropzone/src/DropZone.tsx:40]: Moved the `isVirtualDrag` to RSP DropZone so that we only pass the `aria-labelledby` of the filled message when the dropzone is in both a filled and virtual drag state. 

The reason for including `useDragSession` is because most of the re-rendering is happening in RAC DropZone. So when you call `isVirtualDragging` by itself in RSP DropZone, the value isn’t getting updated. `useDragSession` basically allows us to subscribe to when `beginDragging` is called so the a re-render in RSP DropZone can happen and `isVirtualDragging`’s value can be updated.

File: packages/@react-spectrum/dropzone/src/DropZone.tsx [hunk 4]
```diff
@@ -35,12 +37,17 @@ function DropZone(props: SpectrumDropZoneProps, ref: DOMRef<HTMLDivElement>) {
   let domRef = useDOMRef(ref);
   let messageId = useId();
   let stringFormatter = useLocalizedStringFormatter(intlMessages);
+  let isVirtualDrag = isVirtualDragging(); 
+  let dragSession = DragManager.useDragSession();
+
+  useEffect(() => {
+  }, [dragSession]);
```
Comment (yihuiliao) [inline packages/@react-spectrum/dropzone/src/DropZone.tsx:44]: Moved the `isVirtualDrag` to RSP DropZone so that we only pass the `aria-labelledby` of the filled message when the dropzone is in both a filled and virtual drag state. 

The reason for including `useDragSession` is because most of the re-rendering is happening in RAC DropZone. So when you call `isVirtualDragging` by itself in RSP DropZone, the value isn’t getting updated. `useDragSession` basically allows us to subscribe to when `beginDragging` is called so the a re-render in RSP DropZone can happen and `isVirtualDragging`’s value can be updated.

Comment (snowystinger) [reply packages/@react-spectrum/dropzone/src/DropZone.tsx:44]: why is the useEffect empty?

Comment (yihuiliao) [reply packages/@react-spectrum/dropzone/src/DropZone.tsx:44]: i needed a way to force a re-render on the RSP side when dragSession's state changed, that way the value of isVirtualDragging would update, but there wasn't any other logic to go with it which is why it's empty

Comment (snowystinger) [reply packages/@react-spectrum/dropzone/src/DropZone.tsx:44]: including an empty useEffect won't force a re-render, the only things that cause a render are setState and prop changes
what was the issue you saw if you didn't include it?

Comment (yihuiliao) [reply packages/@react-spectrum/dropzone/src/DropZone.tsx:44]: Ahh I see..so the issue was that when I called `isVirtualDraggging` by itself in RSP DropZone, the value was always false because of most of the re-rendering happens in RAC DropZone and so the value wasn't getting updated. Basically, I just needed a way to for RSP DropZone to know when to update this value.

Comment (LFDanLu) [reply packages/@react-spectrum/dropzone/src/DropZone.tsx:44]: calling the `useDragSession` should be sufficient since that is what will subscribe you to when `startDragging` and `endDragging` are called in DragManager via the setState it sets up (the return type isn't needed, just the setState being called for a rerender)

**Review by LFDanLu — DISMISSED** [review]
LGTM, verified announcements in Chrome/Safari VO

File: packages/@react-spectrum/dropzone/src/DropZone.tsx [hunk 5]
```diff
@@ -12,6 +12,7 @@
 
 import {AriaLabelingProps, DOMProps, DOMRef, StyleProps} from '@react-types/shared';
 import {classNames, SlotProvider, useDOMRef, useStyleProps} from '@react-spectrum/utils';
+import * as DragManager from '@react-aria/dnd/src/DragManager';
```
Comment (devongovett) [inline packages/@react-spectrum/dropzone/src/DropZone.tsx:15]: This import will not work in the built code

**Review by devongovett — CHANGES_REQUESTED** [review]

File: packages/@react-spectrum/dropzone/src/DropZone.tsx (hunk 5 above)
Comment (snowystinger) [reply packages/@react-spectrum/dropzone/src/DropZone.tsx:15]: how do we not have a lint rule for this...

**Review by snowystinger — APPROVED** [review]

**Comment by yihuiliao** [comment]
I've changed this so instead of using DragManger and having the filled message only announce on a virtual drag, it will instead always announce when the dropzone is in a filled state.

**Review by reidbarber — APPROVED** [review]
LGTM
